src/librbd/cache/pwl/ssd/WriteLog.cc

```diff
@@ -975,19 +976,22 @@ void WriteLog<I>::aio_read_data_blocks(
     [this, log_entries, bls, ctx](int r) {
       for (unsigned int i = 0; i < log_entries.size(); i++) {
         bufferlist valid_data_bl;
-        auto length = log_entries[i]->is_write() ? log_entries[i]->write_bytes :
-                                                   log_entries[i]->ws_datalen;
+        auto entry = static_pointer_cast<WriteLogEntry>(log_entries[i]);
+
+        auto length = entry->ram_entry.is_write() ? entry->ram_entry.write_bytes
+                                                  : entry->ram_entry.ws_datalen;
         valid_data_bl.substr_of(*bls[i], 0, length);
         bls[i]->clear();
         bls[i]->append(valid_data_bl);
+        entry->dec_bl_refs();
       }
       ctx->complete(r);
     });
 
   CephContext *cct = m_image_ctx.cct;
   AioTransContext *aio = new AioTransContext(cct, read_ctx);
   for (unsigned int i = 0; i < log_entries.size(); i++) {
-    auto log_entry = log_entries[i];
+    WriteLogCacheEntry *log_entry = &(log_entries[i]->ram_entry);
```

Comment:
I don't think these parentheses are needed either.

Reply:
yes, it is not essential. It has been removed.